lib/resources/aws/aws_security_group.rb

```diff
@@ -9,14 +12,170 @@ class AwsSecurityGroup < Inspec.resource(1)
   supports platform: 'aws'
 
   include AwsSingularResourceMixin
-  attr_reader :description, :group_id, :group_name, :vpc_id
+  attr_reader :description, :group_id, :group_name, :vpc_id, :inbound_rules, :outbound_rules
 
   def to_s
     "EC2 Security Group #{@group_id}"
   end
 
+  def allow_in?(criteria = {})
+    allow(inbound_rules, criteria)
+  end
+  RSpec::Matchers.alias_matcher :allow_in, :be_allow_in
+
+  def allow_out?(criteria = {})
+    allow(outbound_rules, criteria)
+  end
+  RSpec::Matchers.alias_matcher :allow_out, :be_allow_out
+
+  def allow_in_only?(criteria = {})
+    allow_only(inbound_rules, criteria)
+  end
+  RSpec::Matchers.alias_matcher :allow_in_only, :be_allow_in_only
+
+  def allow_out_only?(criteria = {})
+    allow_only(outbound_rules, criteria)
+  end
+  RSpec::Matchers.alias_matcher :allow_out_only, :be_allow_out_only
+
+  def open_to_the_world?
+    allow(inbound_rules, { ipv4_range: '0.0.0.0/0' })
+  end
+
+  def open_to_the_world_on_port?(port)
+    allow(inbound_rules, { ipv4_range: '0.0.0.0/0', port: port })
+  end
+
   private
 
+  def allow_only(rules, criteria)
+    # allow_{in_out}_only require either a single-rule group, or you
+    # to select a rule using position.
+    return false unless rules.count == 1 || criteria.key?(:position)
+    criteria[:exact] = true
+    allow(rules, criteria)
+  end
+
+  def allow(rules, criteria)
+    criteria = allow__check_criteria(criteria)
+    rules = allow__focus_on_position(rules, criteria)
+
+    rules.any? do |rule|
+      matched = true
+      matched &&= allow__match_port(rule, criteria)
+      matched &&= allow__match_protocol(rule, criteria)
+      matched &&= allow__match_ipv4_range(rule, criteria)
+      matched
+    end
+  end
+
+  def allow__check_criteria(raw_criteria)
+    allowed_criteria = [
+      :from_port,
+      :ipv4_range,
+      :port,
+      :position,
+      :protocol,
+      :to_port,
+      :exact, # Internal
+    ]
+    recognized_criteria = {}
+    allowed_criteria.each do |expected_criterion|
+      recognized_criteria[expected_criterion] = raw_criteria.delete(expected_criterion) if raw_criteria.key?(expected_criterion)
```

Comment:
I find conditionals that filters inside of iteration difficult to recognize while reading the code. Perhaps consider filtering first and then building the `recognized_criteria`. Also proposed is a way to remove the empty hash accumulator that is created before the iteration and assign it with the structure that is built from the filtering.

```
supported_criteria = allowed_criteria.find_all { |criterion| expressed_critiera.key?(criterion) }

recognized_criteria = supported_criteria.inject({}) do |recognized,criterion|
  recognized.merge(criterion => raw_criteria.delete(criterion))
end
```

Reply:
I'm writing for maintainability and readability; I appreciate the feedback.
I don't think injection and merge is _more_ clear, however.  Many junior devs struggle with inject.

I'll move the `if` to be a block conditional, instead of an inline conditional, so it stands out.